### PR TITLE
Support for multi version documentation

### LINKED ISSUE #3835: Enable multi-version documentation
### Describe what maintenance you would like added.

Multi-version documentation allows to present various documentations (latest, stable, old) in the same website. In [pyansys/actions](https://actions.docs.pyansys.com), we devised a reusable workflow that only requires:

* The usage of pydata-sphinx-theme (already used by PyVista)
* The usage of GitHub pages

The generation of the documentation artifact is performed by PyVista while [doc-deploy-dev](https://actions.docs.pyansys.com/release/3.0/doc-actions/index.html#doc-deploy-dev-action) and [doc-deploy-stable](https://actions.docs.pyansys.com/release/3.0/doc-actions/index.html#doc-deploy-dev-action) only need to know the name of the documentation artifact to operate with it.

### Links to source code.

* PyAnsys actions: https://actions.docs.pyansys.com
* PyAnsys developers guide: https://dev.docs.pyansys.com/dev/how-to/documenting.html#enabling-multi-version-documentation

### PR DESCRIPTION
**Pull request by AlejandroFernandezLuces**
### Overview
Adds support for multi version documentation by using new actions developed in PyAnsys. With this feature we can have several versions of the documentation in the same domain. Right now, this PR only uploads the multiversion docs to github pages and keeps the current documentation as is, so we don't mess with the current documentation. Once this PR is discussed, we can open another PR to remove the actions related to the old documentation. Closes #3835 

### Details

- Allows both multiversion and single version documentation automatic generation and upload.
- Requires to work with two different folder structure, so we need to upload and download the same documentation twice. This should be solved once we settle on the multiversion docs.

### References
[PyAnsys multi version documentation actions examples](https://actions.docs.pyansys.com/release/3.0/doc-actions/index.html#id1) 

[PyAnsys multi version documentation docs](https://dev.docs.pyansys.com/dev/how-to/documenting.html#enabling-multi-version-documentation)

### CODE REVIEW

File: .github/workflows/docs.yml
Comment (jorgepiloto) [inline .github/workflows/docs.yml:120]: The `working-directory`directive in line 105 only applies for that job step. Hence, for this other step you still need to point to the location of the artifact. Thus:

```suggestion
          path: doc/_build/html/docs-build-multi.tar.gz
```

File: doc/conf.py
Comment (jorgepiloto) [inline doc/conf.py:446]: We usually place this variable close to the "project variables", at the top of the `conf.py` file.

File: .github/workflows/docs.yml
Comment (jorgepiloto) [inline .github/workflows/docs.yml:183]: Let us use the same conditional for deploying:

```suggestion
        if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```

File: .github/workflows/docs.yml
Comment (jorgepiloto) [inline .github/workflows/docs.yml:185]: We are declaring the `CNAME` as an environment variable in the CI/CD pipelines. This way, it is only specified in a single place.

```suggestion
        with:
          cname: ${{ env.DOCUMENTATION_CNAME }}
```

File: .github/workflows/docs.yml
Comment (jorgepiloto) [inline .github/workflows/docs.yml:192]: Same comment about the `CNAME` applies here:

```suggestion
        with:
          cname: ${{ env.DOCUMENTATION_CNAME }}
```

**Review by jorgepiloto — CHANGES_REQUESTED** [review]
This is going in the right direction. Just left some minor comments:

* Declare the `DOCUMENTATION_CNAME` at the very top of the CI/CD pipeline as an environment variable. See other PyAnsys projects using multi-version.

* Make sure you use the `doc-artifact-name` to indicate the `doc-deploy-*` which artifact is the one containing the documentation of the project.

File: .github/workflows/docs.yml
Comment (MaxJPRey) [inline .github/workflows/docs.yml:186]: Reminder, just to not forget to add this environment variable on top of this file.

File: .github/workflows/docs.yml
Comment (jorgepiloto) [inline .github/workflows/docs.yml:25]: This domain should be:

```suggestion
  DOCUMENTATION_CNAME: 'docs.pyvista.org'
```

**Review by jorgepiloto — CHANGES_REQUESTED** [review]
This is looking good, @AlejandroFernandezLuces.

Make sure that the cname in [this versions.json file](https://github.com/pyvista/pyvista/blob/gh-pages/release/versions.json) is the right one. The domain should be `pyvista.org` not `pyvista`.

Also, I think the documentation is failing because pydata-sphinx-theme complains about the `versions.json` file not existing in the current CNAME for the documentation. I bet this is because the latest versions of the `pydata-sphinx-theme` check for this file.

A trick to avoid this may be to place an empty file `versions.json` file in the documentation repositories and before publishing new docs to those repos. This way, documetation build will pass and it won't affect the future CI/CDs.

Once the gh-pages are cluttered, the PyVista maintainers will take the decission to activate those.

**Comment by jorgepiloto** [comment]
My previous approach was a bit complicated. Let's see if we can ignore that link check in the `conf.py`.

**Comment by jorgepiloto** [comment]
Try to add this in the configuartion for the theme: https://pydata-sphinx-theme.readthedocs.io/en/stable/user_guide/version-dropdown.html#configure-switcher-json-url

```python
html_theme_options = {
    # ...
    "check_switcher": False
}
```

**Comment by jorgepiloto** [comment]
It turns out this was the reason for making the docs pass. However, **this failed silently**. Take a look to the logs. [The workflow did not found the artifact](https://github.com/pyvista/pyvista/actions/runs/39974[21](https://github.com/pyvista/pyvista/actions/runs/3997421006/jobs/6858738789#step:17:22)006/jobs/6858738789#step:17:22).

**Comment by AlejandroFernandezLuces** [comment]
> It turns out this was the reason for making the docs pass. However, **this failed silently**. Take a look to the logs. [The workflow did not found the artifact](https://github.com/pyvista/pyvista/actions/runs/39974%5B21%5D(https://github.com/pyvista/pyvista/actions/runs/3997421006/jobs/6858738789#step:17:22)006/jobs/6858738789#step:17:22).

Yes, also I saw earlier that the `gh-pages` branch is empty. There is probably some error on how I managed the paths.

**Comment by banesullivan** [comment]
> Yes, also I saw earlier that the gh-pages branch is empty. There is probably some error on how I managed the paths.

We DO NOT deploy to the github pages branch of this repo as it will blow up the repo size very quickly. Deployed documentation is hosted in separate repos for this reason:

- https://github.com/pyvista/pyvista-docs (releases)
- https://github.com/pyvista/pyvista-docs-dev (`main` branch)

**Comment by AlejandroFernandezLuces** [comment]
> @AlejandroFernandezLuces, I am going to delete the `gh-pages` branch added from this work (see [d0a5a25](https://github.com/pyvista/pyvista/commit/d0a5a25053faac6575268df4ff58ca6dc01c98e1)). Please make sure to push to a separate repository. In fact, please do NOT push to the official repos but to a separate, isolated repo while testing. I made one for you at: https://github.com/pyvista/multiversion-docs-testing

Thanks for creating the repo! We weren't aware that was the reason `pyvista` doesn't have a `gh-pages` itself. We'll do the adaptations needed for deploying in other repos and we'll deploy on the new repo.

**Comment by banesullivan** [comment]
FYI, while deleting the `gh-pages` branch here I noticed it had a shared history with `main`... you may want to make sure the actions your are making create an orphan branch for `gh-pages` as it should definitely not share a history with the source branch. 

FWIW, we currently use https://github.com/peaceiris/actions-gh-pages which I believe does a good job of this... I'd recommend using that action for the deploy step rather than any custom logic

**Comment by banesullivan** [comment]
Indeed, the actions you are making are already composite actions... I'd highly recommend switching this:

https://github.com/pyansys/actions/blob/4ecfff6332d68ed17cf338cc7ea66e77b376e1fd/doc-deploy-stable/action.yml#L153

to use https://github.com/peaceiris/actions-gh-pages

File: .github/workflows/docs.yml
Comment (banesullivan) [inline .github/workflows/docs.yml:107]: nitpick, but why set the working directory to something then perform the operation two levels up? Why not just

```suggestion
        working-directory: ./doc/
        run: tar -zcvf docs-build-multi.tar.gz .
```

File: .github/workflows/docs.yml
Comment (banesullivan) [inline .github/workflows/docs.yml:181]: 😨 ... what's going on here? Again, you definitely should try to use https://github.com/peaceiris/actions-gh-pages to deploy a specific directory and not push the current history

File: .github/workflows/docs.yml
Comment (adeak) [reply .github/workflows/docs.yml:107]: Probably to remove the two enclosing directories from the archive. But I bet `tar` has options to do that.

Comment (banesullivan) [reply .github/workflows/docs.yml:107]: Ohhhhh, good point

Comment (adeak) [reply .github/workflows/docs.yml:107]: Looks like it uses `-C`, just like `make` https://man7.org/linux/man-pages/man1/tar.1.html.

File: .github/workflows/docs.yml
Comment (akaszynski) [reply .github/workflows/docs.yml:181]: Yea, we'll have to clean this up. Assume you're already in a good working state and then proceed.

If I were you, I'd deploy to our [dev-docs](https://dev.pyvista.org/api/index.html) first, get that figured out, and then we can fully replace the documentation. Repo is:
https://github.com/pyvista/pyvista-docs-dev

You guys can play around there, and then once that's working, we can completely overwrite the documentation.

**Review by akaszynski — COMMENTED** [review]
Really excited about the progress with this.

For now, set the target external repo to https://github.com/pyvista/pyvista-docs-dev so we can debug this reasonably.

I think deploying to a separate repository should always be an option, namely being that we don't want to bloat a repo with 100s of MB of documentation. I think the mne docs are a great example with why you'd like to do this. If we're going to track multiple versions, it will get big quickly, especially with something like pyvista where we have a zillion images.

**Comment by banesullivan** [comment]
@akaszynski, I left a comment earlier here that they should use https://github.com/pyvista/multiversion-docs-testing for this PR so as not to affect the documentation used in production.

I want to extra emphasize to please use https://github.com/pyvista/multiversion-docs-testing and NOT https://github.com/pyvista/pyvista-docs-dev until this PR is finished

**Comment by jorgepiloto** [comment]
The great discussion that took place here made us to rethink about the structure and workflows for generating multi-version documentation. Our final thoughts got collected in [this discussion](https://github.com/pyansys/actions/discussions/155).

File: .github/workflows/docs.yml
Comment (tkoyama010) [inline .github/workflows/docs.yml:177]: Is it appropriate to use PyAnsys action here? I believe 100% in the PyAnsys community, but this action is an action for PyAnsys and maybe change in the future. How about forking the action from PyAnsys action and maintaining the action independently in the PyVista community? (If OK, I will fork and maintain.).

Comment (jorgepiloto) [reply .github/workflows/docs.yml:177]: Thanks for raising this question, @tkoyama010. This is a decission you need to agree in PyVista.

We refactored PyAnsys actions according to the feedback of @banesullivan. The only requirements imposed by the documentation deployment actions are:

* Use a theme that inherits from the `pydata-sphinx-theme`
* Apply tags following a `v<Major>.<Minor>.<Patch>` semantic version

I would suggest to go with pyansys/actions for the moment, as they are fully compatible.

If at some point these actions become a limitation, open an issue. If the issue gets discarded, fork and adapt to PyVista needs.

**Comment by jorgepiloto** [comment]
PyAnsys actions v4 is out, see https://actions.docs.pyansys.com/version/stable/

Now, it is possible to deploy to an external repository, desired branch and forcing this branch to be an orphan one.

Pinging here @AlejandroFernandezLuces.

File: .github/workflows/docs.yml
Comment (tkoyama010) [reply .github/workflows/docs.yml:177]: Thank you. We will address the issue when it arises.

File: .github/workflows/docs.yml
Comment (akaszynski) [reply .github/workflows/docs.yml:189]: These should be disabled for now and modified to always deploy to our testing cite even outside of main.

@AlejandroFernandezLuces, what’s the lead time for getting this deployed?

Comment (AlejandroFernandezLuces) [reply .github/workflows/docs.yml:189]: I'll work on this PR this week, I'll try to get it done as soon as possible.

**Comment by AlejandroFernandezLuces** [comment]
The PyVista token does not allow me to deploy to `pyvista/multiversion-docs-testing`: https://github.com/pyvista/pyvista/actions/runs/4511672378/jobs/7945305154

 We probably need a token with permissions to edit that repo, can any of the maintainers check this? @pyvista/maintainers

**Comment by banesullivan** [comment]
@AlejandroFernandezLuces, `${{ secrets.PYVISTA_BOT_TOKEN }}` should be what you need. This is how we currently deploy our docs:

https://github.com/pyvista/pyvista/blob/7f19d5ce90d2d3f7da1e19c630dca3dafd326c94/.github/workflows/docs.yml#L160

**Comment by AlejandroFernandezLuces** [comment]
> @AlejandroFernandezLuces, `${{ secrets.PYVISTA_BOT_TOKEN }}` should be what you need. This is how we currently deploy our docs:
> 
> https://github.com/pyvista/pyvista/blob/7f19d5ce90d2d3f7da1e19c630dca3dafd326c94/.github/workflows/docs.yml#L160

@banesullivan Still not working with the `PYVISTA_BOT_TOKEN`. The error message says: 

`remote: Permission to pyvista/multiversion-docs-testing.git denied to pyvista-bot.` 

I'm not sure why this is happening, do we need to configure something in `pyvista/multiversion-docs-testing` to allow pyvista-bot?

**Comment by banesullivan** [comment]
> I'm not sure why this is happening, do we need to configure something in pyvista/multiversion-docs-testing to allow pyvista-bot?

We synced up offline and this should be fixed now. @AlejandroFernandezLuces, let me know if any further issues

**Comment by AlejandroFernandezLuces** [comment]
I think this is ready on my side. Dev documentation is up in this link: https://pyvista.github.io/multiversion-docs-testing//version/dev/

@banesullivan @jorgepiloto After the review, we can set up a meeting to manually populate the documentation with the stable and previous versions, and then we can merge this to main.

**Review by jorgepiloto — CHANGES_REQUESTED** [review]
Just requested some last changes to this.

File: .github/workflows/docs.yml
Comment (akaszynski) [reply .github/workflows/docs.yml]: Agreed.

**Comment by akaszynski** [comment]
Please provide me with the next steps I need to do to build the docs for the older versions. I image that I'll have to build them locally.

If the docs are linked, let's link them.

Also @jorgepiloto, recommend renaming the action used for this to a non-ansys and more general name. Only "ansys" specific libraries need to be named "ansys". If it's general and can be used in OSS, let's roll with that.

**Comment by AlejandroFernandezLuces** [comment]
> Please provide me with the next steps I need to do to build the docs for the older versions. I image that I'll have to build them locally.
> 
> If the docs are linked, let's link them.
> 
> Also @jorgepiloto, recommend renaming the action used for this to a non-ansys and more general name. Only "ansys" specific libraries need to be named "ansys". If it's general and can be used in OSS, let's roll with that.

To populate the docs with the previous versions, you have to locally build them and then commit them manually to the `versions` folder in multiversion repository. You also have to include the newly added version to `versions.json`.

**Comment by AlejandroFernandezLuces** [comment]
I just manually added stable and dev versions, it seems to be working fine: https://pyvista.github.io/multiversion-docs-testing/version/stable/index.html

If any previous versions need to be included please let me know which versions I should manually add. Once this is merged, newer versions should be automatically deployed.

File: .github/workflows/docs.yml
Comment (akaszynski) [inline .github/workflows/docs.yml:157]:
```suggestion
          repository: 'pyvista/pyvista-docs'
```

File: .github/workflows/docs.yml
Comment (akaszynski) [inline .github/workflows/docs.yml:167]:
```suggestion
          repository: 'pyvista/pyvista-docs'
```

File: .gitignore
Comment (akaszynski) [inline .gitignore:30]:
```suggestion
```

File: doc/source/conf.py
Comment (akaszynski) [inline doc/source/conf.py:282]:
```suggestion
```

File: requirements_docs.txt
Comment (akaszynski) [inline requirements_docs.txt:29]:
```suggestion
```

File: .github/workflows/docs.yml
Comment (akaszynski) [inline .github/workflows/docs.yml:107]:
```suggestion
      - name: Copy metadata
        run: cp doc/source/ads.txt doc/_build/html/
```

**Comment by akaszynski** [comment]
@AlejandroFernandezLuces, thanks for your work here. I think this is finally good to go.

Just so everyone knows, I've already switched out docs to use the multi-version at https://docs.pyvista.org/. This was tested at 0500 UTC, Saturday. It appears that no PyVisters were harmed in the process and the multi-version switcher actually works:
tmp

---

Before making the switch, I was worried about redirects from google. For example:
https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.strip.html

This needs to be redirected to:
https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.PolyDataFilters.strip.html

There isn't really an existing solution for this since the directory structure of our documentation precludes using something like [sphinx-contrib/redirects](https://github.com/sphinx-contrib/redirects). Turns out, there's a fairly easy solution: simply generate one html redirect file for each of the existing ones in `version/stable`. This was done in:
https://github.com/pyvista/pyvista-docs/blob/gh-pages-multiversion/gen-redirect.py

---

### Bonus (because there always is)

We don't do a good job telling Google what it can and can't crawl. We don't even provide a sitemap. While working on the redirects I realized that we should do both. I've added a [sitemap.xml](https://github.com/pyvista/pyvista-docs/blob/gh-pages-multiversion/sitemap.xml), which is generated from [gen-sitemap.py(https://github.com/pyvista/pyvista-docs/blob/gh-pages-multiversion/gen-sitemap.py), as well as [robots.txt](https://github.com/pyvista/pyvista-docs/blob/gh-pages-multiversion/robots.txt).

I've added the sitemap to https://search.google.com/search-console/ in hopes that all pages on Google will go to their redirects.

File: .github/workflows/docs.yml
Comment (tkoyama010) [inline .github/workflows/docs.yml:170]:
```suggestion
```

**Review by banesullivan — APPROVED** [review]
Awesome, awesome work, all! Excited about this! I don't have the bandwidth to thoroughly dive into/review this but have been casually following along, and fell confident in this merging

Though we may want to address this first as it seems like a new issue from the switch: https://github.com/pyvista/pyvista/issues/4283

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Sorry to demoralize you, but I noticed the following thing while using the version switch feature. It works on the top page as shown in this video, but the switch tab does not work on other pages. Is this just my environment?

https://user-images.githubusercontent.com/7513610/232314689-6ade9ac3-e5e0-48b4-a95c-a1c55490dcb6.mp4

**Comment by akaszynski** [comment]
> Sorry to demoralize you, but I noticed the following thing while using the version switch feature. It works on the top page as shown in this video, but the switch tab does not work on other pages. Is this just my environment?

Fixed in [2246a0a](https://github.com/pyvista/pyvista-docs/commit/2246a0aae2da56bb4e7241a637c299cd3afec964). This was already fixed in this branch and backported to the older doc generation. Any new docs generated (after merging this PR) will not have this issue.

**Review by tkoyama010 — APPROVED** [review]
Thanks

**Comment by tkoyama010** [comment]
@AlejandroFernandezLuces Thanks a lot. This PR is a great improvement!

**Comment by AlejandroFernandezLuces** [comment]
Thanks everyone for your feedback! I hope PyVista users will find this useful 😄

**Comment by adeak** [comment]
> I hope PyVista users will find this useful

Considering how dynamically PyVista changes, I think this is super useful to end users. Anyone  constrained to old versions can just look up the corresponding docs 👌

**Comment by jorgepiloto** [comment]
Glad to see this feature got finally merged, @AlejandroFernandezLuces. Congratulations! 🚀

**Pull request merged**
